circuit-construction/src/lib.rs

```diff
@@ -0,0 +1,988 @@
+        let rc = &params.round_constants;
+        let width = PlonkSpongeConstantsKimchi::SPONGE_WIDTH;
+
+        let mut states = vec![input];
+
+        for row in 0..POS_ROWS_PER_HASH {
+            let offset = row * ROUNDS_PER_ROW;
+
+            for i in 0..ROUNDS_PER_ROW {
+                let mut s: Option<Vec<F>> = None;
+                states.push(
+                    (0..3)
+                        .map(|col| {
+                            self.var(|| {
+                                match &s {
+                                    Some(s) => s[col],
+                                    None => {
+                                        // Do one full round on the previous value
+                                        let mut acc = states[states.len() - 1]
+                                            .iter()
+                                            .map(|x| x.val())
+                                            .collect();
+                                        full_round::<F, PlonkSpongeConstantsKimchi>(
+                                            params,
+                                            &mut acc,
+                                            offset + i,
+                                        );
+                                        let res = acc[col];
+                                        s = Some(acc);
+                                        res
+                                    }
+                                }
+                            })
+                        })
+                        .collect(),
+                );
+            }
+
+            self.gate(GateSpec {
+                typ: kimchi::circuits::gate::GateType::Poseidon,
+                c: (0..15)
+                    .map(|i| rc[offset + (i / width)][i % width])
+                    .collect(),
+                row: [
+                    states[offset][0],
+                    states[offset][1],
+                    states[offset][2],
+                    states[offset + 4][0],
+                    states[offset + 4][1],
+                    states[offset + 4][2],
+                    states[offset + 1][0],
+                    states[offset + 1][1],
+                    states[offset + 1][2],
+                    states[offset + 2][0],
+                    states[offset + 2][1],
+                    states[offset + 2][2],
+                    states[offset + 3][0],
+                    states[offset + 3][1],
+                    states[offset + 3][2],
+                ],
+            });
+        }
+
+        let mut final_row = array_init(|_| self.var(|| F::zero()));
+        final_row[0] = states[states.len() - 1][0];
+        final_row[1] = states[states.len() - 1][1];
+        final_row[2] = states[states.len() - 1][2];
+        self.gate(GateSpec {
+            typ: kimchi::circuits::gate::GateType::Zero,
+            c: vec![],
+            row: final_row,
+        });
+
+        states.pop().unwrap()
+    }
+}
```

Comment:
I think these gadgets should live within their respective files, not here. Poseidon, for example, already has a gadget here https://github.com/o1-labs/proof-systems/blob/master/kimchi/src/circuits/polynomials/poseidon.rs#L95

Reply:
Agree-- it's nice to have clean separation and all of a gate's code in a single file or module.  I think the proof-system's hooks into the gate should be as clean as possible and not have to know internal details about the gate.